### ISSUE TITLE
With the same parameters, the video generated using diffusers has poor quality.

### ISSUE THREAD

**Issue by yvmilir** [Wan-Video/Wan2.2#69]
I have repeated the experiment multiple times using the same parameters, but the video quality generated by the Wan2.2 A14B t2v model using diffusers always seems to be lower than that generated using generate.py. Is it because the pipeline in diffusers lacks the guide_scale for the boundry, high-noise, and low-noise stages?

The specific parameter configuration refers to the file wan/configs/wan_t2v_A14B.py

package version: 
diffusers                      0.34.0
modelscope                1.28.1

```python
import torch
import numpy as np
from modelscope import WanPipeline, AutoencoderKLWan
from diffusers.utils import export_to_video, load_image
from diffusers.schedulers.scheduling_unipc_multistep import UniPCMultistepScheduler

dtype = torch.bfloat16
device = "cuda:0"
vae = AutoencoderKLWan.from_pretrained("/mnt/models/Wan2.2-T2V-A14B-Diffusers", subfolder="vae", torch_dtype=torch.float32)
flow_shift = 12.0
pipe = WanPipeline.from_pretrained("/mnt/models//Wan2.2-T2V-A14B-Diffusers", vae=vae, torch_dtype=dtype)
pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config, flow_shift=flow_shift)
pipe.to(device)

height = 480
width = 832

prompt = "Two anthropomorphic cats in comfy boxing gear and bright gloves fight intensely on a spotlighted stage."
negative_prompt = "色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，JPEG压缩残留，丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止不动的画面，杂乱的背景，三条腿，背景人很多，倒着走"
output = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    height=height,
    width=width,
    num_frames=49,
    guidance_scale=4.0,
    num_inference_steps=10,
).frames[0]
import time
timestamp = time.strftime("%Y%m%d-%H%M%S")
export_to_video(output, f"t2v_out_{timestamp}.mp4", fps=30)
``` 

These are the video files I generated, and their frame rates may be different, but the diffusers_demo's quality  is apparently  poor.

Wan2.2 inference code:

https://github.com/user-attachments/assets/8cd8ae47-6703-4438-adc9-d49c51baa455

https://github.com/user-attachments/assets/e7c97394-f2e7-4a89-a3be-a6e3139e1d0b

https://github.com/user-attachments/assets/8f214384-0b33-4523-8d99-6c93ca28b014

diffusers_demo:

https://github.com/user-attachments/assets/f7efeff0-64bc-4b42-a209-ed9857c73126
https://github.com/user-attachments/assets/9acfcc5c-bcac-452a-9ef3-e6fcc8b570b6
https://github.com/user-attachments/assets/bdc338cb-4997-43db-b0d8-d3a995c6982a

**Comment by yvmilir** [comment]
I find that has some recent commits about Wanpipeline in diffusers(main):0.35.0-dev. So I install diffusers(main):0.35.0-dev by source code,  it supports boundry、two-stage guide_scale which conform the diffuser demo by WanAI team in huggingface. 

The video output appears to be fine, at least superior to the diffusers version 0.34.0. So the solution maybe work? 

I would like to know if there will be any future updates to Wan2.2 in diffusers? 

Is the implementation in the current 0.35.0-dev version equivalent to the inference in the current repository inference code (Based on my observation, it seems to be so)?

Looking forward to receiving a reply from the development team. Thanks!

Furthermore, I would like to know if it is necessary to specify a scheduler in pipelines using Wan2.2. I noticed that the example code for diffusers(0.34.0) uses Wan2.1 to specify a scheduler.


video output with diffusers-0.35.0-dev:
https://github.com/user-attachments/assets/67ab35e5-f2d1-4264-bfd8-03dff796568a

```python
import torch
import numpy as np
from modelscope import WanPipeline, AutoencoderKLWan
from diffusers.utils import export_to_video, load_image
from diffusers.schedulers.scheduling_unipc_multistep import UniPCMultistepScheduler

dtype = torch.bfloat16
device = "cuda:0"
vae = AutoencoderKLWan.from_pretrained("/mnt/models/Wan2.2-T2V-A14B-Diffusers", subfolder="vae", torch_dtype=torch.float32)
flow_shift = 12.0
pipe = WanPipeline.from_pretrained("/mnt/models//Wan2.2-T2V-A14B-Diffusers", vae=vae, torch_dtype=dtype)
pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config, flow_shift=flow_shift)
pipe.to(device)

height = 480
width = 832

prompt = "Two anthropomorphic cats in comfy boxing gear and bright gloves fight intensely on a spotlighted stage."
negative_prompt = "色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，JPEG压缩残留，丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止不动的画面，杂乱的背景，三条腿，背景人很多，倒着走"
output = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    height=height,
    width=width,
    num_frames=49,
    guidance_scale=4.0,
    guidance_scale_2=3.0,
    num_inference_steps=10,
).frames[0]
import time
timestamp = time.strftime("%Y%m%d-%H%M%S")
export_to_video(output, f"t2v_out_{timestamp}.mp4", fps=30)
```

**Comment by yvmilir** [comment]
I want to know how to set shift using diffusers? 
[diffusers: UniPCMultistepScheduler set_timesteps function] doesn't have the parameter: shift.

**Comment by MengHao666** [comment]
same problem！ How to solve it？

**Comment by MengHao666** [comment]
any updates？

**Comment by yvmilir** [comment]
> same problem！ How to solve it？

Sorry, I just notice your reply. I had solved it by updating diffusers to 0.35.0-dev. And diffusers has released 0.35.0 and 0.35.1 which support Wan2.2's two-stage processing.  
You can get more information from my another issue:https://github.com/huggingface/diffusers/issues/12094.

**Comment by MengHao666** [comment]
> > same problem！ How to solve it？
> 
> Sorry, I just notice your reply. I had solved it by updating diffusers to 0.35.0-dev. And diffusers has released 0.35.0 and 0.35.1 which support Wan2.2's two-stage processing. You can get more information from my another issue:[huggingface/diffusers#12094](https://github.com/huggingface/diffusers/issues/12094).

I have found the problem.
the problem come from `boundary_ratio` and `sample_shift` parameter.
once you modify the corresponding parameter in diffusers pipeline, the `A14B` model could achieve good performance.

see here:
https://github.com/Wan-Video/Wan2.2/blob/ee56ce852423d780065cd601ac299d0052b0f501/wan/configs/wan_t2v_A14B.py#L36

my usage now:
```


def batch_generate_video(model_id,
                         prompt_file,
                         output_dir):

    import os
    os.makedirs(output_dir, exist_ok=True)
    # prompt
    all_prompts = load_prompts(prompt_file)
    print(f"total {len(all_prompts)} unique prompts!")


    ## prompts##

    # infer config
    negative_prompt = "色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，JPEG压缩残留，丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止不动的画面，杂乱的背景，三条腿，背景人很多，倒着走"
    if model_id == "Wan-AI/Wan2.2-TI2V-5B-Diffusers":
        height = 704
        width = 1280
        num_frames = 121
        num_inference_steps = 50
        guidance_scale = 5.0
        guidance_scale_2 = None
        sample_fps = 24
        sample_shift = 5.0
        boundary_ratio = None
    elif model_id == "Wan-AI/Wan2.2-T2V-A14B-Diffusers":
        # height = 720
        # width = 1280
        height = 480
        width = 832
        num_frames = 81
        num_inference_steps = 40
        guidance_scale = 4.0
        guidance_scale_2 = 3.0
        sample_fps = 16
        sample_shift = 12.0
        boundary_ratio = 0.875
    else:
        raise NotImplementedError


    dtype = torch.bfloat16
    vae = AutoencoderKLWan.from_pretrained(model_id, subfolder="vae", torch_dtype=torch.float32)

    scheduler = UniPCMultistepScheduler.from_pretrained(model_id, subfolder="scheduler", flow_shift=sample_shift)
    pipe = WanPipeline.from_pretrained(model_id,
                                       vae=vae,
                                       scheduler=scheduler,
                                       boundary_ratio=boundary_ratio,
                                       torch_dtype=dtype)




    ### 分布式推理
    from accelerate import Accelerator
    accelerator = Accelerator()
    device = accelerator.device
    pipe = pipe.to(device)

    with torch.no_grad():
        # 关键：分割 sample_info，每个进程获得的都是完整的 sample_info 项
        with accelerator.split_between_processes(sample_info, apply_padding=True) as infos:
            for prompt, prompt_simplified, sample_seed in tqdm(infos):

                # stamp = datetime.now().strftime('%Y%m%d%H%M')
                temp_str = f"{prompt_simplified}_seed{sample_seed}"
                for filename in os.listdir(output_dir):
                    if temp_str in filename:
                        raise FileExistsError(f"{temp_str} already exists in {filename}")

                print("* * "* 10)
                temp_save_path = f"{prompt_simplified}_seed{sample_seed}.mp4"
                final_path = os.path.join(output_dir, f"{prompt_simplified}_seed{sample_seed}.mp4")
                print(f"generating final_path = {final_path}")
                output = pipe(
                    prompt=prompt,
                    negative_prompt=negative_prompt,
                    height=height,
                    width=width,
                    num_frames=num_frames,
                    guidance_scale=guidance_scale,
                    guidance_scale_2=guidance_scale_2,
                    num_inference_steps=num_inference_steps,
                    generator=torch.Generator().manual_seed(sample_seed),  # Set the seed for reproducibility
                ).frames[0]


                export_to_video(output, temp_save_path, fps=sample_fps)
                shutil.move(temp_save_path, final_path)

```

**Comment by INV-WZQ** [comment]
> I have found the problem. the problem come from `boundary_ratio` and `sample_shift` parameter. once you modify the corresponding parameter in diffusers pipeline, the `A14B` model could achieve good performance.
> 
> see here:
> 
> [Wan2.2/wan/configs/wan_t2v_A14B.py](https://github.com/Wan-Video/Wan2.2/blob/ee56ce852423d780065cd601ac299d0052b0f501/wan/configs/wan_t2v_A14B.py#L36)
> 
> Line 36 in [ee56ce8](/Wan-Video/Wan2.2/commit/ee56ce852423d780065cd601ac299d0052b0f501)
> 
>  t2v_A14B.boundary = 0.875 
> my usage now:
> 
> ```
> 
> 
> def batch_generate_video(model_id,
>                          prompt_file,
>                          output_dir):
> 
>     import os
>     os.makedirs(output_dir, exist_ok=True)
>     # prompt
>     all_prompts = load_prompts(prompt_file)
>     print(f"total {len(all_prompts)} unique prompts!")
> 
> 
>     ## prompts##
> 
>     # infer config
>     negative_prompt = "色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，JPEG压缩残留，丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止不动的画面，杂乱的背景，三条腿，背景人很多，倒着走"
>     if model_id == "Wan-AI/Wan2.2-TI2V-5B-Diffusers":
>         height = 704
>         width = 1280
>         num_frames = 121
>         num_inference_steps = 50
>         guidance_scale = 5.0
>         guidance_scale_2 = None
>         sample_fps = 24
>         sample_shift = 5.0
>         boundary_ratio = None
>     elif model_id == "Wan-AI/Wan2.2-T2V-A14B-Diffusers":
>         # height = 720
>         # width = 1280
>         height = 480
>         width = 832
>         num_frames = 81
>         num_inference_steps = 40
>         guidance_scale = 4.0
>         guidance_scale_2 = 3.0
>         sample_fps = 16
>         sample_shift = 12.0
>         boundary_ratio = 0.875
>     else:
>         raise NotImplementedError
> 
> 
>     dtype = torch.bfloat16
>     vae = AutoencoderKLWan.from_pretrained(model_id, subfolder="vae", torch_dtype=torch.float32)
> 
>     scheduler = UniPCMultistepScheduler.from_pretrained(model_id, subfolder="scheduler", flow_shift=sample_shift)
>     pipe = WanPipeline.from_pretrained(model_id,
>                                        vae=vae,
>                                        scheduler=scheduler,
>                                        boundary_ratio=boundary_ratio,
>                                        torch_dtype=dtype)
> 
> 
> 
> 
>     ### 分布式推理
>     from accelerate import Accelerator
>     accelerator = Accelerator()
>     device = accelerator.device
>     pipe = pipe.to(device)
> 
>     with torch.no_grad():
>         # 关键：分割 sample_info，每个进程获得的都是完整的 sample_info 项
>         with accelerator.split_between_processes(sample_info, apply_padding=True) as infos:
>             for prompt, prompt_simplified, sample_seed in tqdm(infos):
> 
>                 # stamp = datetime.now().strftime('%Y%m%d%H%M')
>                 temp_str = f"{prompt_simplified}_seed{sample_seed}"
>                 for filename in os.listdir(output_dir):
>                     if temp_str in filename:
>                         raise FileExistsError(f"{temp_str} already exists in {filename}")
> 
>                 print("* * "* 10)
>                 temp_save_path = f"{prompt_simplified}_seed{sample_seed}.mp4"
>                 final_path = os.path.join(output_dir, f"{prompt_simplified}_seed{sample_seed}.mp4")
>                 print(f"generating final_path = {final_path}")
>                 output = pipe(
>                     prompt=prompt,
>                     negative_prompt=negative_prompt,
>                     height=height,
>                     width=width,
>                     num_frames=num_frames,
>                     guidance_scale=guidance_scale,
>                     guidance_scale_2=guidance_scale_2,
>                     num_inference_steps=num_inference_steps,
>                     generator=torch.Generator().manual_seed(sample_seed),  # Set the seed for reproducibility
>                 ).frames[0]
> 
> 
>                 export_to_video(output, temp_save_path, fps=sample_fps)
>                 shutil.move(temp_save_path, final_path)
> ```


@MengHao666  Hello! I am experiencing an issue with your code. When implementing it on the Wan-AI/Wan2.2-TI2V-5B-Diffusers model, I noticed that the performance is suboptimal. Below, I have included my code, the environment details, and the results. Could you please provide me with some guidance?

### Code
```
import sys
import time
import torch
from transformers import T5TokenizerFast, UMT5EncoderModel
from diffusers.utils import export_to_video
from diffusers import AutoencoderKLWan, WanPipeline, WanTransformer3DModel
from diffusers.schedulers.scheduling_unipc_multistep import UniPCMultistepScheduler


def batch_generate_video(model_id, prompt):
    # infer config
    negative_prompt = "色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，JPEG压缩残留，丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止不动的画面，杂乱的背景，三条腿，背景人很多，倒着走"
    if model_id == "Wan-AI/Wan2.2-TI2V-5B-Diffusers":
        height = 704
        width = 1280
        num_frames = 121
        num_inference_steps = 50
        guidance_scale = 5.0
        guidance_scale_2 = None
        sample_fps = 24
        sample_shift = 5.0
        boundary_ratio = None
    elif model_id == "Wan-AI/Wan2.2-T2V-A14B-Diffusers":
        # height = 720
        # width = 1280
        height = 480
        width = 832
        num_frames = 81
        num_inference_steps = 40
        guidance_scale = 4.0
        guidance_scale_2 = 3.0
        sample_fps = 16
        sample_shift = 12.0
        boundary_ratio = 0.875
    else:
        raise NotImplementedError


    dtype = torch.bfloat16
    vae = AutoencoderKLWan.from_pretrained(model_id, subfolder="vae", torch_dtype=dtype)

    scheduler = UniPCMultistepScheduler.from_pretrained(model_id, subfolder="scheduler", flow_shift=sample_shift)
    pipe = WanPipeline.from_pretrained(model_id,
                                       vae=vae,
                                       scheduler=scheduler,
                                       boundary_ratio=boundary_ratio,
                                       torch_dtype=dtype)

    pipe = pipe.to("cuda")
    output = pipe(
        prompt=prompt,
        negative_prompt=negative_prompt,
        height=height,
        width=width,
        num_frames=9,
        guidance_scale=guidance_scale,
        guidance_scale_2=guidance_scale_2,
        num_inference_steps=num_inference_steps,
    ).frames[0]
    export_to_video(output, "output.mp4", fps=sample_fps)

model_id = "Wan-AI/Wan2.2-TI2V-5B-Diffusers"
prompt = "A cat and a dog baking a cake together in a kitchen. The cat is carefully measuring flour, while the dog is stirring the batter with a wooden spoon. The kitchen is cozy, with sunlight streaming through the window."
batch_generate_video(model_id, prompt)

```

### Environment
```
torch==2.4.1+cu121
diffusers==0.35.1
transformers==4.49.0
```

### Result

https://github.com/user-attachments/assets/1d324002-92b7-4a5e-9369-654c5f9f5aaf

**Comment by MengHao666** [comment]
Sorry for late response. Is it related to the version of transformer or diffusers? I could get good performance with my code.

**Comment by INV-WZQ** [comment]
Thank you very much for your reply. Could you provide the versions of transformers and diffusers you used? This may help me.

**Comment by MengHao666** [comment]
> Thank you very much for your reply. Could you provide the versions of transformers and diffusers you used? This may help me.
following is my setting:

torch260
opencv-python>=4.9.0.80
diffusers>=0.35.1
transformers>=4.49.0
tokenizers>=0.20.3
accelerate>=1.1.1
tqdm
imageio[ffmpeg]
easydict
ftfy
dashscope
imageio-ffmpeg
flash_attn
numpy>=1.23.5,<2
pandas
decord
librosa

**Comment by INV-WZQ** [comment]
Thank you very much! I will try it.
